contrib/babelfishpg_tsql/src/tsql_for/forjson.c

```diff
@@ -265,12 +299,261 @@ tsql_row_to_json(StringInfo state, Datum record, bool include_null_values)
+			}
+		}
+		// Already inserted into existing json object
+		if (found)	
+			continue;
+		
+		// JsonbValue was created in loop and needs to be inserted to the row jsonbValue
+		if (num > 1) {
+			jsonbRow->val.object.nPairs++;
+			jsonbRow->val.object.pairs[jsonbRow->val.object.nPairs - 1] = nestedVal->val.object.pairs[0];
+			continue;
+		}
 
+		// Increment nPairs in the row if it isnt inserted into an already existing json object.
+		jsonbRow->val.object.nPairs++;		
+		colname = parts[0];
+
+		// Allocate memory for key and create it
+		key = palloc(sizeof(JsonbValue));
+		key->type = jbvString;
+		key->val.string.len = strlen(colname);
+		key->val.string.val = pstrdup(colname);
+
+		// Create JsonbPair
+		jsonbPair = palloc(sizeof(JsonbPair));
+		jsonbPair->key = *key;
+		jsonbPair->value = *nestedVal;
+	
+		// Assign it to the JsonbValue Row
+		jsonbRow->val.object.pairs[jsonbRow->val.object.nPairs - 1] = *jsonbPair;
 	}
-	appendStringInfoChar(state, '}');
+
+	// Add the jsonb row to the jsonbArray
+	jsonbArray->val.array.nElems++;
+	jsonbArray->val.array.elems = (JsonbValue *) repalloc(jsonbArray->val.array.elems, sizeof(JsonbValue) * (jsonbArray->val.array.nElems));
+	jsonbArray->val.array.elems[jsonbArray->val.array.nElems - 1] = *jsonbRow;
+
 	ReleaseTupleDesc(tupdesc);
 }
+
+PG_FUNCTION_INFO_V1(tsql_query_to_json_ffunc);
+
+Datum
+tsql_query_to_json_ffunc(PG_FUNCTION_ARGS)
+{
+	forjson_state 	*state;
+	JsonbValue 		*res;
+	Jsonb 			*jsonOut;
+	StringInfo		resStr;
+
+	// Only used if a root_name is given
+	JsonbValue		*root;
+	JsonbValue		*key;
+
+	// Get the processed JsonbValue array
+	state = (forjson_state*) PG_GETARG_POINTER(0);
+	resStr = makeStringInfo();
+
+	if (state->root_name)	{
+
+		// Key jsonBValue to store the root name
+		key = palloc(sizeof(JsonbValue));
+		key->type = jbvString;
+		key->val.string.len = strlen(state->root_name);
+		key->val.string.val = state->root_name;
+		
+		// Root JsonbValue where the key is the root name and value is the processed jsonbVal array
+		root = palloc(sizeof(JsonbValue));
+		root->type = jbvObject;
+		root->val.object.nPairs = 1;
+		root->val.object.pairs = (JsonbPair *) palloc(sizeof(JsonbPair));
+		root->val.object.pairs[0].key = *key;
+		root->val.object.pairs[0].value = *state->jsonbArray;
+
+		// Update the processed jsonbArray
+		state->jsonbArray = root;
+	}
+
+	// Convert JsonbValue to StringInfo for array wrapper check and to return
+	res = state->jsonbArray;
+	jsonOut = JsonbValueToJsonb(res);
+	JsonbToCString(resStr, &jsonOut->root, 0);
+
+	// if without array wrappers is true, remove the array wrappers
+	if (state->without_array_wrapper)	{
+		if (resStr->data[0] == '[')	{
+			resStr->data++;
+			resStr->len--;
+		}
+		if (resStr->data[resStr->len - 1] == ']')	{
+			resStr->data[resStr->len - 1] = '\0';
+			resStr->len--;
+		}
+	}
+	
+	PG_RETURN_TEXT_P(cstring_to_text_with_len(resStr->data, resStr->len));
+}
+
+static char**
+determine_parts(const char* str, int* num)
```

Comment:
For this function, can we just run strtok on copy_str and return that? Is there a reason why we need a nested array?